src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+         */
+        public ContextShard withReads(ArrayList<GATKRead> newReads) {
+            return new ContextShard(this.interval, this.variants, newReads, readContext);
+        }
+
+        /**
+         * creates a new shard, adding the specified read context.
+         * Careful: this call takes ownership of the passed ReadContextData array. So you can't modify it after this call.
+         */
+        public ContextShard withReadContext(ArrayList<ReadContextData> newReadContext) {
+            return new ContextShard(interval, variants, reads, newReadContext);
+        }
+
+        /**
+         * Returns the variants that overlap the query interval, in start-position order.
+         */
+        public ArrayList<Variant> variantsOverlapping(SimpleInterval interval) {
+            return variants.getOverlapping(interval);
+        }
+
+    }
+
+    /**
+     * Fill in reads that start in the given shard, and subshard the output to the requested size.
+     * Signals an error if any read sticks out more than "margin" outside of the big shard.
+     * Optionally filters the reads to only keep the ones that satisfy the given predicate.
+     *
+     * @param bam URL to the reads
+     * @param outputShardSize subsharding size
+     * @param optFilter if specified, only reads that satisfy this will be included.
+     * @return A DoFn that acts as described above.
+     * @throws IOException
+     */
+    public static DoFn<ContextShard,ContextShard> subdivideAndFillReads(String bam, int outputShardSize, int margin, final ReadFilter optFilter) throws IOException {
+        return new DoFnWLog<ContextShard, ContextShard>("subdivideAndFillReads") {
+            private static final long serialVersionUID = 1L;
+            private SamReader reader = null;
+            private Storage.Objects storageClient = null;
+
+            @Override
+            public void startBundle(Context c) throws Exception {
+                super.startBundle(c);
+                if (BucketUtils.isCloudStorageUrl(bam)) {
+                    storageClient = DataflowCommandLineProgram.HellbenderDataflowOptions.Methods.createStorageClient(c.getPipelineOptions().as(DataflowCommandLineProgram.HellbenderDataflowOptions.class));
+                    reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+                } else {
+                    // read from local file (this only makes sense for the direct runner)
+                    reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
+                }
+            }
+
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                ContextShard shard = c.element();
+                ArrayList<SimpleInterval> ints =new ArrayList<>();
+                ints.add(shard.interval);
+                List<SimpleInterval> subshards = ShardedReadsDatasource.cutToShards(ints, outputShardSize);
```

Comment:
Aha, so `cutToShards()` at least is still used -- in that case, my earlier comment about standalone unit tests for that method still applies.